File: Monitor-python/event-monitor.py
import time
import traceback
from web3 import Web3
import os

class Event_Filter():
    def read_ABI_from_file(self, contract_name):
        try:
            current_dir = os.path.dirname(os.path.abspath(__file__))
            file = open(f'{current_dir}/abi/{contract_name}', 'r')
            return file.read()
        except:
            return None

    def read_last_block(self, file_path):
        try:
            with open(file_path, 'r') as file:
                return int(file.read().strip())
        except:
            return w3.eth.blockNumber

    def save_last_block(self, file_path, block_number):
        with open(file_path, 'w+') as file:
            file.write(str(block_number))

    def handle_migration(self, event):
        print("emit Migration event")
        user = event['args']['user']
        token_id = event['args']['tokenId']
        amount = event['args']['amount']
        migration_time = event['args']['timestamp']
        transaction_hash = event['transactionHash'].hex()
        print(f"user address: {user}, token ID: {token_id}, amount: {amount}, timestamp: {migration_time}, transaction hash: {transaction_hash}")

    def monitor_events(self, w3, contract_address, network, block_size, contract_abi, from_block, current_block):
        contract = w3.eth.contract(address=contract_address, abi=contract_abi)
        if from_block == 0 or from_block > current_block:
            from_block = current_block - block_size  # recent 1500 blocks

        while from_block < current_block:
            to_block = min(from_block + (block_size - 1), current_block)
            print(f'Now processing {network} blocks {from_block} to {to_block}...')

            event_filter_Migration = contract.events.Migration.createFilter(fromBlock=from_block, toBlock=to_block)

            for event in event_filter_Migration.get_all_entries():
                self.handle_migration(event)

            from_block = to_block + 1

eventFilter = Event_Filter()
RPC = {
       'Base_testnet': 'https://84532.rpc.thirdweb.com/'
        # add more RPC URLs
        # 'Polygon_mainnet': 'https://polygon-mainnet.g.alchemy.com/v2/THISISKEY',
        # 'Avalanche_mainnet': 'https://avax-mainnet.g.alchemy.com/v2/THISISKEY',
        # 'opBNB_mainnet': 'https://bsc-dataseed.binance.org/',
       }

# NETWORK = ['Polygon', 'Base', 'Avalanche', 'opBNB']
NETWORK = ['Base']
# NETTYPE = 'mainnet'
NETTYPE = 'testnet'
BLOCK_SIZE = 1500
CONTRACT_ADDRESS = '0x1dF4DB808FD43A867A087D38c01e804E3b3e6f81'

EVENT_FILTER_INTERVAL = 60
while True:
    for NET in NETWORK:
        try:
            BLOCK_FILE = f'./block/{NET}_{NETTYPE}.txt'
            ABI_FILE = 'NFT_Migration'
            abi = eventFilter.read_ABI_from_file(ABI_FILE)

            # Web3 인스턴스 생성 및 Ethereum 노드에 연결
            rpc_url = RPC[f'{NET}_{NETTYPE}']
            w3 = Web3(Web3.HTTPProvider(rpc_url))
            # fromBlock 읽기
            from_block = eventFilter.read_last_block(BLOCK_FILE)
            # 최신 블록 번호 가져오기
            current_block = w3.eth.blockNumber
            # 이벤트 모니터링
            eventFilter.monitor_events(w3, CONTRACT_ADDRESS, NET, BLOCK_SIZE, abi, from_block, current_block)
            # 마지막 조회한 블록 번호 저장
            eventFilter.save_last_block(BLOCK_FILE, current_block)
        except:
            traceback.print_exc()
            print(f'Failed to get data from {NET} {NETTYPE}.')

    time.sleep(EVENT_FILTER_INTERVAL)

File: Monitor-nodejs/event-monitor.js
const fs = require('fs').promises;
const path = require('path');
const Web3 = require('web3');

class EventFilter {
    async readABIFromFile(contractName) {
        try {
            const currentDir = path.dirname(require.main.filename);
            const content = await fs.readFile(`${currentDir}/abi/${contractName}`, 'utf8');
            return content;
        } catch (error) {
            console.error('Error reading ABI file:', error);
            return null;
        }
    }

    async readLastBlock(filePath) {
        try {
            const content = await fs.readFile(filePath, 'utf8');
            return parseInt(content.trim());
        } catch (error) {
            console.error('Error reading last block:', error);
            return null;
        }
    }

    async saveLastBlock(filePath, blockNumber) {
        try {
            await fs.writeFile(filePath, blockNumber.toString());
        } catch (error) {
            console.error('Error saving last block:', error);
        }
    }

    handleMigration(event) {
        console.log("emit Migration event");
        const user = event.returnValues.user;
        const tokenId = event.returnValues.tokenId;
        const amount = event.returnValues.amount;
        const migrationTime = event.returnValues.timestamp;
        const transactionHash = event.transactionHash;
        console.log(`user address: ${user}, token ID: ${tokenId}, amount: ${amount}, timestamp: ${migrationTime}, transaction hash: ${transactionHash}`);
    }

    async monitorEvents(w3, contractAddress, network, blockSize, contractAbi, fromBlock, currentBlock) {
        const contract = new w3.eth.Contract(JSON.parse(contractAbi), contractAddress);

        if (fromBlock === 0 || fromBlock > currentBlock) {
            fromBlock = currentBlock - blockSize; // recent 1500 blocks
        }

        while (fromBlock < currentBlock) {
            const toBlock = Math.min(fromBlock + (blockSize - 1), currentBlock);
            console.log(`Now processing ${network} blocks ${fromBlock} to ${toBlock}...`);

            try {
                const events = await contract.getPastEvents('Migration', {
                    fromBlock: fromBlock,
                    toBlock: toBlock
                });

                for (const event of events) {
                    this.handleMigration(event);
                }
            } catch (error) {
                console.error(`Error getting events for blocks ${fromBlock} to ${toBlock}:`, error);
            }

            fromBlock = toBlock + 1;
        }
    }
}

const RPC = {
    'Base_testnet': 'https://84532.rpc.thirdweb.com/'
    // add more RPC URLs as needed
    // 'Polygon_mainnet': 'https://polygon-mainnet.g.alchemy.com/v2/THISISKEY',
    // 'Avalanche_mainnet': 'https://avax-mainnet.g.alchemy.com/v2/THISISKEY',
    // 'opBNB_mainnet': 'https://bsc-dataseed.binance.org/',
};

const NETWORK = ['Base'];
const NETTYPE = 'testnet';
const BLOCK_SIZE = 1500;
const CONTRACT_ADDRESS = '0x1dF4DB808FD43A867A087D38c01e804E3b3e6f81';
const EVENT_FILTER_INTERVAL = 60000; // milliseconds

async function main() {
    const eventFilter = new EventFilter();

    while (true) {
        for (const NET of NETWORK) {
            try {
                const BLOCK_FILE = path.join('./block', `${NET}_${NETTYPE}.txt`);
                const ABI_FILE = 'NFT_Migration';
                const abi = await eventFilter.readABIFromFile(ABI_FILE);

                if (!abi) {
                    console.error(`Failed to read ABI for ${NET}`);
                    continue;
                }

                const rpcUrl = RPC[`${NET}_${NETTYPE}`];
                const w3 = new Web3(new Web3.providers.HttpProvider(rpcUrl));

                let fromBlock = await eventFilter.readLastBlock(BLOCK_FILE);

                const currentBlock = await w3.eth.getBlockNumber();

                await eventFilter.monitorEvents(
                    w3,
                    CONTRACT_ADDRESS,
                    NET,
                    BLOCK_SIZE,
                    abi,
                    fromBlock,
                    currentBlock
                );

                await eventFilter.saveLastBlock(BLOCK_FILE, currentBlock);
            } catch (error) {
                console.error(`Failed to get data from ${NET} ${NETTYPE}:`, error);
            }
        }

        await new Promise(resolve => setTimeout(resolve, EVENT_FILTER_INTERVAL));
    }
}

main().catch(console.error);